pkg/stdlib/sql.go

```diff
@@ -0,0 +1,377 @@
+/*
+Copyright 2021 CodeNotary, Inc. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package stdlib
+
+import (
+	"context"
+	"database/sql"
+	"database/sql/driver"
+	"fmt"
+	"github.com/codenotary/immudb/pkg/api/schema"
+	"github.com/codenotary/immudb/pkg/client"
+	"google.golang.org/grpc/metadata"
+	"io"
+	"math"
+	"reflect"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+)
+
+var immuDriver *Driver
+
+func init() {
+	immuDriver = &Driver{
+		configs: make(map[string]*Conn),
+	}
+	sql.Register("immudb", immuDriver)
+}
+
+type DBOption func(*Connector)
+
+type ConnConfig struct {
+	client.Options
+}
+
+func OpenDB(cliOpts *client.Options) *sql.DB {
+	c := Connector{
+		cliOptions: cliOpts,
+		driver:     immuDriver,
+	}
+
+	return sql.OpenDB(c)
+}
+
+type Connector struct {
+	cliOptions *client.Options
+	driver     *Driver
+}
+
+// RegisterConnConfig registers a ConnConfig and returns the connection string to use with Open.
+func RegisterConnConfig(c *Conn) string {
+	return immuDriver.registerConnConfig(c)
+}
+
+// UnregisterConnConfig removes the ConnConfig registration for connStr.
+func UnregisterConnConfig(connStr string) {
+	immuDriver.unregisterConnConfig(connStr)
+}
+
+// Connect implement driver.Connector interface
+func (c Connector) Connect(ctx context.Context) (driver.Conn, error) {
+	conn, err := client.NewImmuClient(c.cliOptions)
+	if err != nil {
+		return nil, err
+	}
+	lr, err := conn.Login(ctx, []byte(c.cliOptions.Username), []byte(c.cliOptions.Password))
+	if err != nil {
+		return nil, err
+	}
+
+	md := metadata.Pairs("authorization", lr.Token)
+	ctx = metadata.NewOutgoingContext(ctx, md)
+
+	resp, err := conn.UseDatabase(ctx, &schema.Database{DatabaseName: c.cliOptions.Database})
+	if err != nil {
+		return nil, err
+	}
+	cn := &Conn{
+		conn:    conn,
+		options: c.cliOptions,
+		Token:   resp.Token,
+	}
+	c.driver.registerConnConfig(cn)
+
+	return &Conn{conn: conn, options: c.cliOptions, Token: resp.Token}, nil
+}
+
+// Driver implement driver.Connector interface
+func (c Connector) Driver() driver.Driver {
+	return c.driver
+}
+
+type Driver struct {
+	configMutex sync.Mutex
+	configs     map[string]*Conn
+	sequence    int
+}
+
+func (d *Driver) Open(name string) (driver.Conn, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), 60*time.Second)
```

Comment:
I'm not sure we have to register those here (correct me if i'm wrong), but even official docs state (https://pkg.go.dev/database/sql#Open):

> Most users will open a database via a driver-specific connection helper function that returns a *DB.

Maybe we could skip that part in the first version of the driver?

Then it should ideally parse strings like 'user:passs@host:port'

Reply:
As discussed offline, this will be needed by gorm migrations